### DISCUSSION TITLE
lftp issues on Windows

### DISCUSSION THREAD

**Discussion by gvanem** [getpelican/pelican#3145] [Q&A]
I discovered Pelican some days ago and was quite pleased.
That until I tried to find a suitable upload program for Windows.
The generated `Makefile` had a `lftp` command that failed to work for me.
I did install a Cygwin-x64 version of `lftp.exe`:
```
LFTP | Version 4.9.2 | Copyright (c) 1996-2020 Alexander V. Lukyanov
...
Libraries used: Expat 2.5.0, idn2 2.3.4, libiconv 1.17, OpenSSL 1.1.1u  30 May 2023, Readline 8.2, zlib 1.2.13
```

The issues I had:
 * The server to upload to (*one.com* in Denmark) has both an IPv4 and IPv6 address. But my IPv6 connectivity is 
 broken thanks to my ISP Telenor and my crappy router.  But the command `set dns:order inet` forced IPv4 only.
* Cygwin's `lftp` doesn't understand Windows paths like `f:/gv/Python/whatever`. 
Had to convert it to `/cygdrive/f/gv/Python/whatever`.
* Using the default TLS settings in `lftp` caused the 1st upload to never finish. Related to [this](https://github.com/lavv17/lftp/issues/346)?
  But a `set ftp:ssl-allow false` fixed that.
 * And finally, the uploaded files were all with `CHMOD 750`. 
 Thus my web-client (Chrome) got a `403` response on my Pelican site.

So I ended up with these make-rules that worked for me:
```
...
OUTPUTDIR     = $(BASEDIR)/output
OUTPUTDIR_CYG = /cygdrive/$(subst :,,$(OUTPUTDIR))

LFTP_CMD = set ftp:passive-mode true ; \
           set ftp:ssl-allow false   ; \
           set dns:order inet        ; \
           mirror -R $(OUTPUTDIR_CYG) $(FTP_TARGET_DIR) ; \
           quit

ftp_upload: publish run_chmod
	lftp ftp://$(FTP_USER_PWD)@$(FTP_HOST) -e "$(LFTP_CMD)"

run_chmod:
	cd $(OUTPUTDIR) ; \
	chmod --recursive --verbose 755 *
```

I mean, how hard should be to publish a Web-site? Anybody else on Windows had issues like these?

**Comment by justinmayer** [comment]
In my experience, it's not hard to publish a web-site. Then again, I haven't used FTP in… decades? If your host offers SSH access, I think you might be better served by Rsync, which I have found to work very well.

**Comment by gvanem** [comment]
>  If your host offers SSH access, 

It does, but I find ssh equally clumsy as these I've looked into:
* WinSCP and it's [synchronize](https://winscp.net/eng/docs/scriptcommand_synchronize)  command.
*  [SiteCopy](https://github.com/notroj/sitecopy/) which even crashes on minor issues. 
* Have not looked at Rsync; will install the Cygwin port of it.

*PS*. If I use relative paths in my makefile, the horrid `OUTPUTDIR_CYG = /cygdrive/$(subst :,,$(OUTPUTDIR))` is not needed.

**Comment by justinmayer** [comment]
As I do not use Windows, I am ill-equipped to provide guidance in this area. Perhaps someone else can provide Windows-specific suggestions.
